### PR TITLE
[READY]Circuitry upgrades

### PR DESCRIPTION
**Pull request by arsserpentarium**
HO-HO-HO!Merry X-mas!
And I present to you some fixes and upgrades in circuitry:
1.make plant module kill vines
2.xenobio scanner
3.heater module
4.damage for bot and machines(?)
5.improved pathfinder circuit
6.NTnet scaner(takes ref to thing.returns adress of NTnet adress if there is one)
7.fix work of arithmetic pins
8.object scaner. attack assembly with something on help intent and obtain it's ref.
9.Reagent juicer and grinder. When ACCount will remove possibility to drain fruits via injector, it could help.
10.fixed some runtimes in juicing.
image
image
reupload.sorry.
Sorry for pulling claw. Mainteiners asked me to do it in separate PR. So it will be right after this PR.

### CODE REVIEW

**Comment by KomradeSpectre** [comment]
I am so sorry for this merge conflict, my addition that was sitting for a week got merged, and it had more icons than your pull does, hence why this happened.

**Comment by Tlaltecuhtli** [comment]
3.what do you mean by smoker fixed?
8. ???

**Comment by Tlaltecuhtli** [comment]
4*

**Comment by arsserpentarium** [comment]
3.fixed smoker ref pin+raised cooldown.
8.Do you know,what is ntnet?

**Comment by Tlaltecuhtli** [comment]
does it find the ntnet adress inside other drones?

**Comment by arsserpentarium** [comment]
Yes,it is.

**Comment by Tlaltecuhtli** [comment]
what happens if i have more than one ntnet inside

**Comment by arsserpentarium** [comment]
It gives first founded.

**Comment by arsserpentarium** [comment]
Technically it means that module,which wad inserted earlier.

**Comment by arsserpentarium** [comment]
What do you think about idea to make reagent grinder relatevely slightly damage mobs and gib corpses?

**Comment by arsserpentarium** [comment]
also,please, put on this PR "fix" label, because half of my changes is fixes.

**Comment by ninjanomnom** [comment]
image
You fucked it up again

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 1]
```diff
@@ -29,6 +30,8 @@
 	.=..()
 	START_PROCESSING(SScircuit, src)
 	materials[MAT_METAL] = round((max_complexity + max_components) / 4) * SScircuit.cost_multiplier
+	if (!armor)
+		armor = list(melee = 50, bullet = 70, laser = 70, energy = 100, bomb = 10, bio = 100, rad = 100, fire = 0, acid = 0)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/core/assemblies.dm]: set this in `/obj/item/device/electronic_assembly`

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 2]
```diff
@@ -326,8 +329,14 @@
 		interact(user)
 		return TRUE
 	else
+		if(user.a_intent == INTENT_HELP)
+			for(var/obj/item/integrated_circuit/input/objscaner/S in assembly_components)
+				if(S.check_then_do_work())
+					S.scan(I,user)
+					return TRUE
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/core/assemblies.dm]: Just make a generic hook for integrated circuits instead of looking for objscanners in particular.

File: code/modules/hydroponics/grown.dm [hunk 3]
```diff
@@ -146,15 +146,15 @@
 
 /obj/item/reagent_containers/food/snacks/grown/on_grind()
 	var/nutriment = reagents.get_reagent_amount("nutriment")
-	if(grind_results.len)
+	if(length(grind_results))
```
Comment (ninjanomnom) [inline code/modules/hydroponics/grown.dm]: Revert this, more thorough testing earlier revealed `.len` is faster than `length` like we originally thought.

File: code/game/machinery/magnet.dm [hunk 4]
```diff
@@ -21,7 +21,7 @@
 	var/code = 0 // frequency code, they should be different unless you have a group of magnets working together or something
 	var/turf/center // the center of magnetic attraction
 	var/on = FALSE
-	var/pulling = 0
+	var/pulling1 = 0
```
Comment (ninjanomnom) [inline code/game/machinery/magnet.dm]: Don't do bad name changes like this

File: code/modules/integrated_electronics/subtypes/arithmetic.dm [hunk 5]
```diff
@@ -30,10 +30,9 @@
 /obj/item/integrated_circuit/arithmetic/addition/do_work()
 	var/result = 0
 	for(var/k in 1 to inputs.len)
-		var/datum/integrated_io/I = inputs[k]
-		I.pull_data()
-		if(isnum(I.data))
-			result = result + I.data
+		var/I = get_pin_data(IC_INPUT, k)
+		if(isnum(I))
+			result = result + I
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/arithmetic.dm]: `result += I`

File: code/modules/integrated_electronics/subtypes/arithmetic.dm [hunk 6]
```diff
@@ -56,13 +55,10 @@
...
-		var/datum/integrated_io/I = inputs[k]
-		if(I == A)
-			continue
-		I.pull_data()
-		if(isnum(I.data))
-			result = result - I.data
+	for(var/k in 2 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(isnum(I))
+			result = result - I
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/arithmetic.dm]: `result -= I`

File: code/modules/integrated_electronics/subtypes/arithmetic.dm [hunk 7]
```diff
@@ -85,13 +81,10 @@
...
-		var/datum/integrated_io/I = inputs[k]
-		if(I == A)
-			continue
-		I.pull_data()
-		if(isnum(I.data))
-			result = result * I.data
+	for(var/k in 2 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(isnum(I))
+			result = result * I
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/arithmetic.dm]: `result *= I`

File: code/modules/integrated_electronics/subtypes/arithmetic.dm [hunk 8]
```diff
@@ -114,13 +107,12 @@
...
-		if(I == A)
-			continue
-		I.pull_data()
-		if(isnum(I.data) && I.data != 0) //No runtimes here.
-			result = result / I.data
+
+	for(var/k in 2 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(isnum(I) && (I != 0))
+			result = result / I
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/arithmetic.dm]: `result /= I`

File: code/modules/integrated_electronics/subtypes/arithmetic.dm [hunk 9]
```diff
@@ -228,12 +220,11 @@
...
-	for(var/k in 1 to inputs.len)
-		var/datum/integrated_io/I = inputs[k]
-		I.pull_data()
-		if(isnum(I.data))
+	for(var/k in 2 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(isnum(I))
 			inputs_used++
-			result = result + I.data
+			result = result + I
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/arithmetic.dm]: `result += I`

File: code/modules/integrated_electronics/subtypes/arithmetic.dm [hunk 10]
```diff
@@ -288,11 +279,10 @@
...
 	var/result = 0
-	for(var/k in 1 to inputs.len)
-		var/datum/integrated_io/I = inputs[k]
-		I.pull_data()
-		if(isnum(I.data))
-			result = sqrt(I.data)
+	for(var/k in 2 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(isnum(I))
+			result = sqrt(I)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/arithmetic.dm]: Result is reset each iteration of the loop and previous values mean nothing so you should step backwards instead
```
for(var/k in inputs.len to 2 step -1)
   var/I = get_pin_data(IC_INPUT, k)
   if(isnum(I)
      result = sqrt(I)
      break
```

File: config/game_options.txt [hunk 11]
```diff
@@ -492,4 +492,4 @@ MICE_ROUNDSTART 10
 #EMERGENCY_SHUTTLE_AUTOCALL_THRESHOLD 0.2
 
 ## Determines if players are allowed to print integrated circuits, uncomment to allow.
-#IC_PRINTING
\ No newline at end of file
+IC_PRINTING
```
Comment (ninjanomnom) [inline config/game_options.txt]: revert this

**Review by ninjanomnom — DISMISSED** [review]

File: code/modules/integrated_electronics/core/assemblies.dm (hunk 2 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/core/assemblies.dm]: Generic hook?where can I see,what is it?

File: code/modules/integrated_electronics/subtypes/arithmetic.dm (hunk 10 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/arithmetic.dm]: I missed there +=

**Comment by arsserpentarium** [comment]
Looks like it is ready.

File: code/game/objects/objs.dm [hunk 12]
```diff
@@ -63,6 +66,85 @@
 		visible_message("<span class='danger'>[src] shatters into a million pieces!</span>")
 		qdel(src)
 
+/obj/proc/start_pulling(atom/movable/AM,gs)
```
Comment (ninjanomnom) [inline code/game/objects/objs.dm]: If we need pulling on obj and mob, it should be an atom proc.

This *should* be it's own pr.

File: code/modules/integrated_electronics/core/integrated_circuit.dm [hunk 13]
```diff
@@ -58,6 +58,15 @@ a creative player the means to solve many problems.  Circuits are held inside an
 /obj/item/integrated_circuit/proc/any_examine(mob/user)
 	return
 
+/obj/item/integrated_circuit/proc/attackby_react(var/atom/movable/A,mob/user)
+	return
+
+/obj/item/integrated_circuit/proc/sence(var/atom/movable/A,mob/user)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/core/integrated_circuit.dm]: sense

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
You need to atomize your prs more and not just throw together code changes for various things into one pr. The pulling change is significant enough to merit it's own refactor.

File: code/game/objects/objs.dm (hunk 12 above)
Comment (arsserpentarium) [reply code/game/objects/objs.dm]: I've tryed. And it worse. It broken lights. And mob still need its own pulling proc.
Also this proc is slightly different on mobs and objs. Sad,but not worked.

**Comment by arsserpentarium** [comment]
Well... it was all things for circuitry upgrade.

**Comment by awesine** [comment]
downvoted purely because you're increasing the smoke cooldown

it's almost impossible to actually use circuits right and nerfing on one of the things 90% of the playerbase couldn't figure out is bad

**Comment by arsserpentarium** [comment]
@awesine 
awesine,smoker was on 30 sec cooldown on Polaris. I have set it to 1 sec by mistake.I am trying to avoid nerfs at all costs.

File: code/game/objects/objs.dm (hunk 12 above)
Comment (ninjanomnom) [reply code/game/objects/objs.dm]: You'll probably have to do this change first in another pr *then* come back to this one

Comment (arsserpentarium) [reply code/game/objects/objs.dm]: I've separated claw from everything else. I'll make it as another PR.

**Comment by arsserpentarium** [comment]
looks like requested changes done.

**Comment by arsserpentarium** [comment]
Honestly,I really don't want exclude grabbing claw from PR. I wanted to make drone chariot,ask admins for santa's clothes and start to ride across station, giving neat circuitry presents to crew.

**Comment by awesine** [comment]
@arsserpentarium 
ok and

**Comment by arsserpentarium** [comment]
@awesine , let's have a deal. If you will bring me 3 possible implementation for fast smoker, besides of annoy people, I'll lower cooldown.

File: code/game/machinery/magnet.dm [hunk 14]
```diff
@@ -21,7 +21,7 @@
 	var/code = 0 // frequency code, they should be different unless you have a group of magnets working together or something
 	var/turf/center // the center of magnetic attraction
 	var/on = FALSE
-	var/pulling = 0
+	var/magneting = 0
```
Comment (ninjanomnom) [inline code/game/machinery/magnet.dm]: You can reverse this change since the reason this was done is gone from this pr now.

File: code/game/objects/objs.dm [hunk 15]
```diff
@@ -28,6 +28,9 @@
 	var/list/unique_reskin //List of options to reskin.
 	var/dangerous_possession = FALSE	//Admin possession yes/no
 
+	var/atom/movable/pulling = null
+	var/grab_state = 0
```
Comment (ninjanomnom) [inline code/game/objects/objs.dm]: Same as above

File: code/modules/integrated_electronics/subtypes/reagents.dm [hunk 16]
```diff
@@ -318,6 +323,86 @@
...
+
+/obj/item/integrated_circuit/reagent/storage/grinder/do_work()
+	if(reagents.total_volume >= reagents.maximum_volume)
+		activate_pin(3)
+		return FALSE
+	var/obj/item/I = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
+	if(istype(I))
+		if(check_target(I))
+			if(I.on_grind(src) != -1)
+				if(I.grind_results)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/reagents.dm]: `if(istype(I) && check_target(I) && I.on_grind(src) != -1 && I.grind_results)`

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 17]
```diff
@@ -686,6 +730,49 @@
 	spawn_flags = IC_SPAWN_DEFAULT|IC_SPAWN_RESEARCH
 	power_draw_per_use = 120
 
+/obj/item/integrated_circuit/input/sensor/ranged/sense_remote(var/atom/A,mob/user)
+	if(!user)
+		if(user.client)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/input.dm]: there's no user, how can you check client of a non-existent user?

File: code/modules/integrated_electronics/subtypes/manipulation.dm [hunk 18]
```diff
@@ -247,9 +249,15 @@
 /obj/item/integrated_circuit/manipulation/plant_module/do_work()
 	..()
 	var/turf/T = get_turf(src)
-	var/obj/machinery/hydroponics/AM = get_pin_data_as_type(IC_INPUT, 1, /obj/machinery/hydroponics)
+	var/obj/OM = get_pin_data_as_type(IC_INPUT, 1, /obj)
+	if(istype(OM,/obj/structure/spacevine))
+		if(get_pin_data(IC_INPUT, 2)==2)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/manipulation.dm]: save on indentation
`if(istype(OM,/obj/structure/spacevine) && get_pin_data(IC_INPUT, 2) == 2)`

File: code/modules/integrated_electronics/subtypes/reagents.dm [hunk 19]
```diff
@@ -318,6 +323,86 @@
...
+
+/obj/item/integrated_circuit/reagent/storage/juicer/do_work()
+	if(reagents.total_volume >= reagents.maximum_volume)
+		activate_pin(3)
+		return FALSE
+	var/obj/item/I = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
+	if(istype(I))
+		if(check_target(I))
+			if(I.juice_results)
+				if(I.on_juice() != -1)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/reagents.dm]: `if(istype(I) && check_target(I) && I.juice_results && I.on_juice() != -1)`

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/subtypes/input.dm (hunk 17 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/input.dm]: there is user.It is called with afterattack.

File: code/modules/integrated_electronics/subtypes/reagents.dm (hunk 19 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/reagents.dm]: if on_juice will be called without juice_results ,it can cause fuck up.

**Comment by epochayur** [comment]
you're doing god's work, arsser

**Comment by arsserpentarium** [comment]
done.

**Comment by arsserpentarium** [comment]
Thank you, epochayur.

File: code/modules/integrated_electronics/subtypes/input.dm (hunk 17 above)
Comment (ninjanomnom) [reply code/modules/integrated_electronics/subtypes/input.dm]: `if(!user)`

File: code/modules/integrated_electronics/subtypes/reagents.dm [hunk 20]
```diff
@@ -318,6 +323,82 @@
...
+	complexity = 16
+	spawn_flags = IC_SPAWN_RESEARCH
+
+/obj/item/integrated_circuit/reagent/storage/juicer/do_work()
+	if(reagents.total_volume >= reagents.maximum_volume)
+		activate_pin(3)
+		return FALSE
+	var/obj/item/I = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
+	if(istype(I)&&check_target(I)&&(I.juice_results))
+		if(I.on_juice() != -1)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/reagents.dm]: As I said before:
`if(istype(I) && check_target(I) && I.juice_results && I.on_juice() != -1)`
These kinds of conditionals short circuit if it knows earlier that the check as a whole is going to fail.
`check_target()` wont be run if the `istype()` fails. `I.on_juice()` wont be run if there is no `I.juice_results`

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by arsserpentarium** [comment]
fixed

File: code/modules/integrated_electronics/subtypes/reagents.dm (hunk 20 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/reagents.dm]: needed knowlege

**Comment by ninjanomnom** [comment]
There's still one requested change

**Comment by arsserpentarium** [comment]
uhh.missed it.sorry.

**Comment by arsserpentarium** [comment]
looks like everything is ok.

**Comment by arsserpentarium** [comment]
by the way. @awesine
I've reverted cooldown to 1 sec.Kevinz nerf is enought for balancing

**Comment by arsserpentarium** [comment]
all requested changes has been done

File: code/modules/hydroponics/grown.dm [hunk 21]
```diff
@@ -146,19 +146,21 @@
 
 /obj/item/reagent_containers/food/snacks/grown/on_grind()
 	var/nutriment = reagents.get_reagent_amount("nutriment")
-	if(grind_results.len)
-		for(var/i in 1 to grind_results.len)
-			grind_results[grind_results[i]] = nutriment
-		reagents.del_reagent("nutriment")
-		reagents.del_reagent("vitamin")
+	if(grind_results)
+		if(grind_results.len)
```
Comment (ninjanomnom) [inline code/modules/hydroponics/grown.dm]: `if(length(grind_results))` does both checks in one go

File: code/modules/hydroponics/grown.dm [hunk 22]
```diff
@@ -146,19 +146,21 @@
...
 
 /obj/item/reagent_containers/food/snacks/grown/on_juice()
 	var/nutriment = reagents.get_reagent_amount("nutriment")
-	if(juice_results.len)
-		for(var/i in 1 to juice_results.len)
-			juice_results[juice_results[i]] = nutriment
-		reagents.del_reagent("nutriment")
-		reagents.del_reagent("vitamin")
+	if(juice_results)
+		if(juice_results.len)
```
Comment (ninjanomnom) [inline code/modules/hydroponics/grown.dm]: same as above

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 23]
```diff
@@ -326,8 +317,13 @@
 		interact(user)
 		return TRUE
 	else
+		if(user.a_intent == INTENT_HELP)
+			for(var/obj/item/integrated_circuit/input/objscaner/S in assembly_components)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/core/assemblies.dm]: end the type path at /input
this should be able to support other kinds of input circuits if need be

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 24]
```diff
@@ -274,19 +274,10 @@
...
-						continue
-			else
-				continue
-		S.set_pin_data(IC_OUTPUT, 1, WEAKREF(target))
-		S.check_then_do_work()
-		S.scan(target)
+		if(proximity)
+			S.sense(target,user)
+		else
+			S.sense_remote(target,user)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/core/assemblies.dm]: make a generic hook like attackby_react below

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/core/assemblies.dm (hunk 23 above)
Comment (ninjanomnom) [reply code/modules/integrated_electronics/core/assemblies.dm]: and have it work with all intents while you're at it, just pass the intent in as an argument

**Review by ninjanomnom — APPROVED** [review]
